File: app/src/main/java/com/sharpdroid/registroelettronico/Fragments/FragmentLessons.java
package com.sharpdroid.registroelettronico.Fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.sharpdroid.registroelettronico.API.SpiaggiariApiClient;
import com.sharpdroid.registroelettronico.Activities.MainActivity;
import com.sharpdroid.registroelettronico.Adapters.AllLessonsAdapter;
import com.sharpdroid.registroelettronico.Databases.SubjectsDB;
import com.sharpdroid.registroelettronico.Interfaces.API.Lesson;
import com.sharpdroid.registroelettronico.R;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;

import static com.sharpdroid.registroelettronico.Utils.Metodi.getProfessorOfThisSubject;
import static com.sharpdroid.registroelettronico.Utils.Metodi.getSubjectName;
import static com.sharpdroid.registroelettronico.Utils.Metodi.isNetworkAvailable;


public class FragmentLessons extends Fragment implements SwipeRefreshLayout.OnRefreshListener {

    private final static String TAG = "Lessons";
    static int code;
    @BindView(R.id.recycler)
    RecyclerView mRecyclerView;
    @BindView(R.id.coordinator_layout)
    CoordinatorLayout mCoordinatorLayout;
    @BindView(R.id.swiperefresh)
    SwipeRefreshLayout mSwipeRefreshLayout;
    AllLessonsAdapter mRVAdapter;
    SubjectsDB db;

    public FragmentLessons() {
    }

    public static FragmentLessons newInstance(int code) {
        FragmentLessons fragment = new FragmentLessons();
        Bundle b = new Bundle();
        b.putInt("code", code);
        fragment.setArguments(b);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            code = getArguments().getInt("code");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_recycler_refresh_scrollbar, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        db = new SubjectsDB(getContext());
        MainActivity activity = (MainActivity) getActivity();

        activity.setTitle(getSubjectName(db.getSubject(code)));

        mRVAdapter = new AllLessonsAdapter(getContext());
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerView.setAdapter(mRVAdapter);

        mSwipeRefreshLayout.setRefreshing(true);
        mSwipeRefreshLayout.setOnRefreshListener(this);
        mSwipeRefreshLayout.setColorSchemeResources(
                R.color.bluematerial,
                R.color.redmaterial,
                R.color.greenmaterial,
                R.color.orangematerial);
        bindLessonsCache();

        UpdateLessons();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        db.close();
    }

    @Override
    public void onRefresh() {
        UpdateLessons();
    }

    private void addLessons(List<Lesson> lessons) {
        if (!lessons.isEmpty()) {
            mRVAdapter.clear();
            mRVAdapter.addAll(lessons);
        }
    }

    private void bindLessonsCache() {
        addLessons(db.getLessons(code));
    }

    private void UpdateLessons() {
        if (!isNetworkAvailable(getContext())) {
            Snackbar.make(mCoordinatorLayout, R.string.nointernet, Snackbar.LENGTH_LONG).show();
            mSwipeRefreshLayout.setRefreshing(false);
        } else {
            for (Integer prof : db.getProfessorCodes(code)) {
                new SpiaggiariApiClient(getContext())
                        .getLessons(code, String.valueOf(prof))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(lessons -> {
                            //update subjectsDB
                            db.removeLessons(code);
                            db.addLessons(code, prof, lessons);
                            db.addProfessor(code, prof, getProfessorOfThisSubject(lessons));
                            bindLessonsCache();

                            mSwipeRefreshLayout.setRefreshing(false);
                        }, error -> {
                            error.printStackTrace();
                            mSwipeRefreshLayout.setRefreshing(false);
                        });
            }
        }
    }
}

File: app/src/main/java/com/sharpdroid/registroelettronico/fragments/FragmentLessons.kt
package com.sharpdroid.registroelettronico.fragments

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.ContentViewEvent
import com.sharpdroid.registroelettronico.BuildConfig
import com.sharpdroid.registroelettronico.NotificationManager
import com.sharpdroid.registroelettronico.R
import com.sharpdroid.registroelettronico.adapters.AllLessonsAdapter
import com.sharpdroid.registroelettronico.database.pojos.LessonMini
import com.sharpdroid.registroelettronico.database.viewModels.LessonsViewModel
import com.sharpdroid.registroelettronico.utils.EventType
import com.sharpdroid.registroelettronico.utils.Metodi.updateLessons
import com.sharpdroid.registroelettronico.views.EmptyFragment
import kotlinx.android.synthetic.main.coordinator_swipe_recycler.view.*
import kotlinx.android.synthetic.main.fragment_recycler_refresh_scrollbar.*


class FragmentLessons : Fragment(), SwipeRefreshLayout.OnRefreshListener, NotificationManager.NotificationReceiver {
    override fun didReceiveNotification(code: EventType, args: Array<in Any>) {
        when (code) {
            EventType.UPDATE_LESSONS_START -> {
                swiperefresh.isRefreshing = true
            }
            EventType.UPDATE_LESSONS_OK,
            EventType.UPDATE_LESSONS_KO -> {
                swiperefresh.isRefreshing = false
            }
            else -> { // Ignore
            }
        }
    }

    private lateinit var mRVAdapter: AllLessonsAdapter
    private lateinit var emptyHolder: EmptyFragment

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val layout = inflater.inflate(R.layout.fragment_recycler_refresh_scrollbar, container, false)
        emptyHolder = EmptyFragment(context)
        emptyHolder.visibility = View.GONE
        emptyHolder.setTextAndDrawable("Nessuna lezione", R.drawable.ic_view_agenda)
        layout.coordinator_layout.addView(emptyHolder)
        return layout
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        NotificationManager.instance.addObserver(this, EventType.UPDATE_LESSONS_KO, EventType.UPDATE_LESSONS_OK, EventType.UPDATE_LESSONS_START)

        mRVAdapter = AllLessonsAdapter()
        recycler.layoutManager = LinearLayoutManager(context)
        recycler.adapter = mRVAdapter

        swiperefresh.setOnRefreshListener(this)
        swiperefresh.setColorSchemeResources(
                R.color.bluematerial,
                R.color.redmaterial,
                R.color.greenmaterial,
                R.color.orangematerial)


        ViewModelProviders.of(activity)[LessonsViewModel::class.java].selected.observe(this, Observer {
            activity.title = it?.getSubjectName().orEmpty()
            addLessons(it?.lessons.orEmpty())
        })

        //onRefresh()
        if (!BuildConfig.DEBUG)
            Answers.getInstance().logContentView(ContentViewEvent().putContentId("Lezioni").putContentType("Lezioni"))
    }

    override fun onRefresh() {
        updateLessons(context)
    }

    private fun addLessons(lessons: List<LessonMini>) {
        mRVAdapter.clear()
        mRVAdapter.addAll(lessons.sortedByDescending { it.mDate })

        emptyHolder.visibility = if (lessons.isEmpty()) View.VISIBLE else View.GONE
    }

    override fun onDestroyView() {
        super.onDestroyView()
        NotificationManager.instance.removeObserver(this, EventType.UPDATE_LESSONS_KO, EventType.UPDATE_LESSONS_OK, EventType.UPDATE_LESSONS_START)
    }
}
